File: pc02/pc02_01/c/pc02_01.c
/**
 * Jolly jumpers
 * pc110201
 * uva10038
 */

#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 3000

int is_jolly(int array[MAX_SIZE], int size) {
	int seen_differences[MAX_SIZE - 1] = {0};

	for (int i = 0; i < size - 1; i++) {
		int difference = abs(array[i + 1] - array[i]);

		if (difference < 1 || difference > size - 1) {
			return 0;
		}

		if (seen_differences[difference - 1] == 1) {
			return 0;
		}

		seen_differences[difference - 1] = 1;
	}

	return 1;
}

int main(void) {
	int array[MAX_SIZE];
	int size;

	while (scanf("%d", &size) != EOF) {
		for (int i = 0; i < size; i++) {
			scanf("%d", &array[i]);
		}

		if (is_jolly(array, size)) {
			printf("Jolly\n");
		} else {
			printf("Not jolly\n");
		}
	}

	return 0;
}

File: pc02/pc02_01/python/pc02_01.py
"""
Jolly jumpers
pc110201
uva10038
"""

import sys


def is_jolly(array):
    """
    Check if array is jolly (all consecutive differences are distinct and in [1, n-1])
    """
    size = len(array)
    seen_differences = [False] * (size - 1)

    for i in range(size - 1):
        difference = abs(array[i + 1] - array[i])

        if difference < 1 or difference > size - 1:
            return False

        if seen_differences[difference - 1]:
            return False

        seen_differences[difference - 1] = True

    return True


def main():
    for line in sys.stdin:
        parts = line.strip().split()
        if not parts:
            continue

        size = int(parts[0])
        if size == 0:
            continue

        array = list(map(int, parts[1:]))

        if is_jolly(array):
            print("Jolly")
        else:
            print("Not jolly")


if __name__ == '__main__':
    main()
